### PR TITLE
Added connection template to filestructure

### PR DESCRIPTION
**Pull request by JHocevar**
Updates file structure with template connection module.
* Most of the code comes from @Rhijke 's [fork](https://github.com/Rhijke/datajoint-core)
* Updated python module and test file to look more like [datajoint-python](https://github.com/datajoint/datajoint-python)
* Added cargo workspace to build cargo modules from root
* Python connection module will look for compiled rust library in /target/debug/corelib

### CODE REVIEW

File: datajoint-python/datajoint/connection.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import os, sys, ctypes
+from ctypes import c_uint32, c_char_p, c_uint8, c_void_p, Structure, POINTER, c_bool
+
+class ConnectionS(Structure):
```
Comment (jackson-nestelroad) [inline datajoint-python/datajoint/connection.py:4]: Probably should make this name slightly more detailed (ConnectionStruct) if it has to be included like this.

File: datajoint-python/datajoint/connection.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+prefix = {'win32': ''}.get(sys.platform, 'lib')
+extension = {'darwin': '.dylib', 'win32': '.dll'}.get(sys.platform, '.so')
+dirname = os.path.dirname(__file__)
+library_file = os.path.join(dirname + '/../../target/debug/' + prefix + 'corelib' + extension)
+lib = ctypes.cdll.LoadLibrary(library_file)
+
+lib.connection_new.restype = POINTER(ConnectionS)
+
+lib.connection_free.argtypes = (POINTER(ConnectionS), )
```
Comment (jackson-nestelroad) [inline datajoint-python/datajoint/connection.py:15]: Extra comma?

Comment (JHocevar) [reply datajoint-python/datajoint/connection.py:15]: Trailing comma is python notation to create a tuple of 1 item, and the argtypes property needs a tuple.

File: datajoint-python/datajoint/connection.py (hunk 1 above)
Comment (JHocevar) [reply datajoint-python/datajoint/connection.py:4]: I agree, I'll update it to ConnectionStruct for now, and we can discuss cleaner options later.

File: packages/datajoint-core/src/connection/mod.rs [hunk 3]
```diff
@@ -1,3 +1,26 @@
...
+
+impl Connection {
+    pub fn new() -> Connection {
+        Connection {
+        }
+    }
+
+    // Support for MySQL and Postgres database servers
+    // No dynamic database configuration (hard-coded into the code for now perhaps at compile time)
+    pub fn connect(&mut self, host: &str, user: &str, password: &str, reset: bool , use_tls: bool) {
```
[the remark this answers is not in the document]
Comment (JHocevar) [reply packages/datajoint-core/src/connection/mod.rs]: Agreed, updated

File: packages/datajoint-core/src/connection/mod.rs [hunk 4]
```diff
@@ -1,3 +1,26 @@
...
+        println!("Connected to database: Host: {} User: {} Password: {} Reset: {} use_TLS: {} ", host, user, password, reset, use_tls);
+    }
+
+    // Provide a utility in the lib to receive a Generic SQL query and
+    //      execute against a relational database server
+    // No placeholder arguments (hard-coded into the queries for now)
+    pub fn query(&self, query: &str) -> u32 {
```
[the remark this answers is not in the document]
Comment (JHocevar) [reply packages/datajoint-core/src/connection/mod.rs]: agreed, updated

**Review by Rhijke — DISMISSED** [review]

File: packages/datajoint-core-ffi-c/src/lib.rs [hunk 5]
```diff
@@ -1,7 +1,74 @@
...
+pub extern "C" fn connection_connect(ptr: *mut Connection) {
+    let database = unsafe {
+        assert!(!ptr.is_null());
+        &mut *ptr
+    };
+    database.connect();
 }
+
+#[no_mangle]
+pub extern "C" fn connection_query(
```
Comment (jackson-nestelroad) [inline packages/datajoint-core-ffi-c/src/lib.rs:60]: Should also name this to raw_query to match the Rust function name.

**Review by jackson-nestelroad — CHANGES_REQUESTED** [review]

File: packages/datajoint-core-ffi-c/src/lib.rs (hunk 5 above)
Comment (JHocevar) [reply packages/datajoint-core-ffi-c/src/lib.rs]: fixed

**Review by guzman-raphael — CHANGES_REQUESTED** [review]
@JHocevar Thanks for the PR and nice work! :muscle: 

Nice you've gone through the trouble of accessing `cffi` vs `ctypes` in Python, looks like we can perhaps close this one in favor of #11. Seems like that `cffi` is the best way to go since it reduces the burden of declaration since we can reuse the `cbindgen` generated header.

**Pull request closed without merging**